core/src/main/java/com/linecorp/armeria/common/metric/ServiceNaming.java

```diff
@@ -0,0 +1,122 @@
+ *  version 2.0 (the "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+package com.linecorp.armeria.common.metric;
+
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.common.logging.RequestOnlyLog;
+import com.linecorp.armeria.common.util.Unwrappable;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.ServiceConfig;
+
+/**
+ * Generates the default name of a {@link com.linecorp.armeria.server.Service} from its {@link ServiceConfig}.
+ *
+ * @see RequestOnlyLog#serviceName()
+ * @see com.linecorp.armeria.server.HttpService
+ * @see com.linecorp.armeria.server.RpcService
+ */
+@FunctionalInterface
+public interface ServiceNaming {
+
+    /**
+     * Returns the {@link ServiceNaming} that always returns the given hard-coded service name.
+     */
+    static ServiceNaming of(String defaultServiceName) {
+        requireNonNull(defaultServiceName, "defaultServiceName");
+        return cfg -> defaultServiceName;
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the simple name of innermost class from the given service.
+     * It is supposed to have a class name without a period. e.g. {@code HelloService}.
+     *
+     * @see Class#getSimpleName()
+     */
+    static ServiceNaming simpleTypeName() {
+        return cfg -> {
+            Unwrappable unwrappable = cfg.service();
+            for (;;) {
+                final Unwrappable delegate = unwrappable.unwrap();
+                if (delegate != unwrappable) {
+                    unwrappable = delegate;
+                    continue;
+                }
+                return delegate.getClass().getSimpleName();
+            }
+        };
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the full name of innermost class from the given service.
+     * It will be fully qualified name including the package name separated by a period.
+     * e.g. {@code com.foo.HelloService}.
+     *
+     * @see Class#getName()
+     */
+    static ServiceNaming fullyQualifiedName() {
+        return cfg -> {
+            Unwrappable unwrappable = cfg.service();
+            for (;;) {
+                final Unwrappable delegate = unwrappable.unwrap();
+                if (delegate != unwrappable) {
+                    unwrappable = delegate;
+                    continue;
+                }
+                return delegate.getClass().getName();
+            }
+        };
+    }
```

Comment:
If this is applied by default, this service naming breaks changes. Because this is not compatible with the existing service naming for RPC services.
https://github.com/line/armeria/blob/03becb62cd04991ca1488f6a3b9bcd228aeb09d7/core/src/main/java/com/linecorp/armeria/common/logging/RequestOnlyLog.java#L183-L188

Reply:
Exactly. It is also related your other comments, https://github.com/line/armeria/pull/3366#discussion_r601989466. I want to discuss it. For me, in case of server metrics, fully qualified service name makes sense instead of RPC service names. However, it causes breaking changes as you said. If the compatibility should be kept, one of following changes is expected
1) change the timing of calling `ServiceNaming.convert()`
2) does not determine `defaultServiceName` at constructor in case of RPC service. In other words, rollback `ServiceConfig.defaultServiceName` to nullable.
WDYT?

Reply:
How about this:

- Make `ServiceConfig.defaultServiceName()` always return `null` and deprecate it.
- Change `ServiceNaming` to require `ServiceRequestContext` instead of `ServiceConfig`.
- Don't calculate service name in the beginning. Calculate the service name in `DefaultRequestLog` like we did before.

This should simplify things a little bit more. What do you think?

Reply:
I agree. That should work for sure. I will address with below comments. Thanks again @ikhoon @trustin